### DISCUSSION TITLE
Web2print page settings best practices?

### DISCUSSION THREAD

**Discussion by NiklasBr** [pimcore/pimcore#13453] [Q&A / Support]
For some web2print products there is a need to have different margins or different headers/footers or no footer on some pages and so on.

Since it seems the Web To Print Processor only loads the web2print configuration once for each run, I wonder if there are any suggestions or best practices to adhere to when generating pages like this in the same PDF:

![image](https://user-images.githubusercontent.com/279826/197985391-454885db-5162-4f20-9107-9723431c3261.png)

*and* like this: 

![image](https://user-images.githubusercontent.com/279826/197985543-322766fd-626f-4578-9919-7851480613a9.png)

**Comment by kingjia90** [comment]
It seems that we just need to do some improvement there, as the case you mentioned might not be covered.

I have noticed that for Headless Chrome, the `$config` parameter is unused. 
What if we make it be possible to change custom settings at document level(like the screenshow below) by merging custom values in `jobConfigFile` (which is a json as well) with the global ones (in web2print settings, acting as default value) or something like that?

https://github.com/pimcore/pimcore/blob/aff1e27db9d3774ac8b5b266cc8c21faca1f2b23/lib/Web2Print/Processor/HeadlessChrome.php#L36-L73

The following is what was already possible with PDFReactor:
![image](https://user-images.githubusercontent.com/6014195/199018129-28275103-84f0-4945-91a3-6a9f0beb3316.png)

**Reply by NiklasBr** [reply]
Sounds good. Anything that can resolve this issue would be nice,

By the way did you manage to install Chrome Headless on the Demo instance so that it is easier to reproduce and report issues?

**Reply by kingjia90** [reply]
Cool, i try to see if it won't take too much effort as i feel there are already all the foundation to make it quickly work as described. 
If not, i'll open an improvement issue and see if it can be prioritized. 

About the headless chrome in demo, partially [yes](https://github.com/pimcore/demo/pull/329), but unfortunately there were still some problems with node 18 and wasn't possible to merge that 😞

**Reply by kingjia90** [reply]
I've just opened a PR  https://github.com/pimcore/pimcore/pull/13493.
Probably not the best approach (eg. doesn't store settings in document but in json/psf, needs to hardcode the available options etc..), but at least it quickly fixes by "using" what was already there for PdfReactor and without introducing any big changes or refactoring.

Hope it help and looking forward for feedbacks

**Reply by NiklasBr** [reply]
Thanks! I'll test it today and see what happens

**Reply by NiklasBr** [reply]
@kingjia90 

Having tried to understand web2print even further I have some questions:

1. What is the expected way for one user to access the settings made by another to a document? These appear to be saved with the User ID, but it is not clear why and what use case there is. It creates confusion.
2. What s the expected way to load processingoptions during a `PrintContainer` PDF generation? E.g. a user sets some processing options on different `PrintPage`s, then opens the `PrintContainer` and clicks the "Generate PDF" button, how come the different options does not appear to be loaded?
3. Why are the config options (those from `abstract public function getProcessingOptions()`) not made available in any Events during PrintPage save?

**Reply by kingjia90** [reply]
I am flattered but i am not the right person to answer these questions.  
My only guess is that the current implementation was made for a specific use case that required that workflow, which coincidentally covers most of the generic usages but didn't go more in depth for more advanced usages, and since not seeing much issues, improvement request nor PRs in that direction, it was just left to be used as is.

**Reply by NiklasBr** [reply]
Okay, though it seems like such an specific thing to connect settings such as headers, footers, margins and things like that to a specific user. There must be a reason, but I can't find any documentation to support why.

**Comment by NiklasBr** [comment]
Trying to look into a related web2print issue it seems like the Demo Instance can no longer generate PDF:s from Printcontainers. Is this known?

![image](https://user-images.githubusercontent.com/279826/199956797-759599cb-c558-4b89-b593-973851731dd8.png)
